ktor-features/ktor-jrpc/src/io/ktor/jrpc/routers/JrpcRouter.kt

```diff
@@ -0,0 +1,191 @@
+                            JrpcResponse(requestId, error = JrpcError.INVALID_PARAMS)
+                        } else {
+                            JrpcResponse(requestId, handler(context, methodParams))
+                        }
+                    }
+                }
+            } catch (je: JrpcException) {
+                JrpcResponse(requestId, error = JrpcError("${je.message}:\r\n${
+                if (JrpcSupport.printStackTraces) stackTraceToString(je) else "stacktrace hidden"
+                }", je.code))
+            } catch (cce: ClassCastException) {
+                JrpcResponse(requestId, error = JrpcError("error calling ${jrpcRequest.method} ${cce.localizedMessage}. Handler can not be cast to PipelineContext<Unit, ApplicationCall>.(Any) -> Any?", JrpcError.CODE_INTERNAL_ERROR))
+            } catch (e: Exception) {
+                JrpcResponse(requestId, error = JrpcError("error calling ${jrpcRequest.method} ${e.localizedMessage}", JrpcError.CODE_INTERNAL_ERROR))
+            }
+        } else {
+            JrpcResponse(requestId, error = JrpcError.METHOD_NOT_FOUND)
+        }
+    }
+
+    private fun stackTraceToString(throwable: Throwable): String {
+        var result = ""
+        StringWriter().use { sw ->
+            PrintWriter(sw).use { pw ->
+                throwable.printStackTrace(pw)
+            }
+            result = sw.toString()
+        }
+        return result
+    }
+
+    /**
+     * Add new JRPC-method for handling. The param will converted to requested type if it possible.
+     * The result will be added to result of JrpcResponse
+     * If you need to throw from method, you may use [JrpcException]
+     * @param T type of input argument to which will be converted "params" key from JRPC-request. Pass Unit if you don't have any input params, pass Any? if you want to convert if by yourself
+     * @param method name of JRPC method
+     * @param block method body with requested receiver T
+     * @throws RuntimeException if method with the same name exists in handlers
+     */
+    @ContextDsl
+    inline fun <reified T : Any> method(method: String, noinline block: PipelineContext<Unit, ApplicationCall>.(T) -> Any?) {
+        if (methodAlreadySet(method)) throw RuntimeException("Method with name \"$method\" already exists in handlers")
+        else addMethod(method, JrpcRouteItem(block, T::class.java))
+    }
+
+    /**
+     * If handler parse JrpcRequest properly, will called before handle any JRPC-method
+     * Usable for checks, headers, etc
+     * @param block - block of code that will be executed. If block return JrpcResponse, the other JRPC methods will not be handled, the method returns immediately with that JrpcResponse. Note that params will be not parsed at the moment.
+     * @throws RuntimeException if called twice
+     */
+    @ContextDsl
+    fun beforeJrpc(block: PipelineContext<Unit, ApplicationCall>.(JrpcRequest) -> JrpcResponse?) {
```

Comment:
It looks like we need a pipeline here instead of before/after hooks. Similar to `ApplicationCallPipeline` we can use `JrpcCallPipeline` with appropriate phases